controlplane/kubeadm/internal/control_plane.go

```diff
@@ -87,23 +87,21 @@ func (c *ControlPlane) EtcdImageData() (string, string) {
 // MachinesNeedingUpgrade return a list of machines that need to be upgraded.
 func (c *ControlPlane) MachinesNeedingUpgrade() FilterableMachineCollection {
 	now := metav1.Now()
-	var requireUpgrade FilterableMachineCollection
 	if c.KCP.Spec.UpgradeAfter != nil && c.KCP.Spec.UpgradeAfter.Before(&now) {
```

Comment:
Not necessary: We can drop hash check twice if we do the following: 

```
	requireUpgrade := c.Machines.Filter(
		machinefilters.Not(machinefilters.MatchesConfigurationHash(hash.Compute(&c.KCP.Spec))))
	if c.KCP.Spec.UpgradeAfter != nil && c.KCP.Spec.UpgradeAfter.Before(&now) {
		requireUpgrade = requireUpgrade.AnyFilter(machinefilters.OlderThan(c.KCP.Spec.UpgradeAfter))
	}
	return requireUpgrade
```

Reply:
I like your suggestion better, but I'm going to keep the logic as is for now